enzyme/Enzyme/MLIR/Dialect/Ops.cpp

```diff
@@ -191,3 +191,23 @@ LogicalResult BatchOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
 
   return success();
 }
+
+//===----------------------------------------------------------------------===//
+// BroadcastOp
+//===----------------------------------------------------------------------===//
+
+void BroadcastOp::build(OpBuilder &builder, OperationState &result, Value input, llvm::SmallVector<int64_t> shape) {
+  auto shapeAttr = builder.getDenseI64ArrayAttr(shape);
+  RankedTensorType output;
+  // TODO: support things other than scalars and ranked tensors, maybe reuse getShadowType here?
```

Comment:
yeah long term we should probably just use https://github.com/EnzymeAD/Enzyme/blob/534a28595a9bf1f41d4a8222615f00e36611dc0f/enzyme/Enzyme/MLIR/Interfaces/AutoDiffTypeInterface.td#L59

so essentially do

auto ty = input.getType();
for (auto s : reverse(shape)) {
  ty = ty.cast<AutodiffTypeInterface>().getShadowType(s)
}

Reply:
but for now this is fine